File: sdks/python/taskito/dashboard/oauth/providers.py
"""Concrete provider implementations: Google, GitHub, generic OIDC.

Every provider satisfies :class:`OAuthProvider`. The split between
``exchange_code`` (network IO + claim normalisation) and
``check_allowlist`` (pure-data permission check) is deliberate so tests
can drive either path in isolation.

Providers depend on :class:`HttpClient`, a structural Protocol over
the small subset of ``requests.Session`` they actually use (one ``get``
method). Production code passes a ``requests.Session``; tests pass an
in-memory stub. The Protocol keeps the provider layer framework-free
and test-friendly without runtime ``cast`` calls at either boundary.
"""

from __future__ import annotations

import time
from typing import TYPE_CHECKING, Any, Protocol
from urllib.parse import urlencode

import requests
from authlib.integrations.requests_client import OAuth2Session
from joserfc import jwt
from joserfc.errors import JoseError
from joserfc.jwk import KeySet

from taskito.dashboard.oauth.identity import (
    AllowlistDenied,
    IdentityFetchError,
    ProviderIdentity,
)

if TYPE_CHECKING:
    from taskito.dashboard.oauth.config import (
        GitHubConfig,
        GoogleConfig,
        OIDCConfig,
    )


class HttpResponse(Protocol):
    """Minimal response shape a provider consumes from its HTTP client."""

    status_code: int
    text: str

    def json(self) -> Any: ...

    def raise_for_status(self) -> None: ...


class HttpClient(Protocol):
    """Minimal HTTP client shape — ``requests.Session`` satisfies this."""

    def get(
        self,
        url: str,
        *,
        headers: dict[str, str] | None = ...,
        timeout: float = ...,
    ) -> HttpResponse: ...


GOOGLE_DISCOVERY_URL = "https://accounts.google.com/.well-known/openid-configuration"
GITHUB_AUTH_URL = "https://github.com/login/oauth/authorize"
GITHUB_TOKEN_URL = "https://github.com/login/oauth/access_token"
GITHUB_API_BASE = "https://api.github.com"

_HTTP_TIMEOUT = 10.0


def _email_domain(email: str | None) -> str | None:
    if not email or "@" not in email:
        return None
    return email.rsplit("@", 1)[-1].lower()


def _audience_matches(aud: Any, client_id: str) -> bool:
    if isinstance(aud, str):
        return aud == client_id
    if isinstance(aud, list):
        return client_id in aud
    return False


# ── OIDC provider (shared logic for Google + generic OIDC) ─────────────


class _OIDCProviderBase:
    """Shared OIDC machinery: discovery, JWKS caching, ID-token decoding."""

    slot: str
    label: str
    type: str
    client_id: str
    client_secret: str
    discovery_url: str
    scope: str = "openid email profile"

    def __init__(self, *, http: HttpClient | None = None) -> None:
        self._http = http or requests.Session()
        self._discovery: dict[str, Any] | None = None
        self._jwks: dict[str, Any] | None = None

    # Sub-classes override / extend `_extra_auth_params` to add hints.
    def _extra_auth_params(self) -> dict[str, str]:
        return {}

    def _get_discovery(self) -> dict[str, Any]:
        if self._discovery is None:
            resp = self._http.get(self.discovery_url, timeout=_HTTP_TIMEOUT)
            resp.raise_for_status()
            self._discovery = resp.json()
        return self._discovery

    def _get_jwks(self) -> dict[str, Any]:
        if self._jwks is None:
            resp = self._http.get(self._get_discovery()["jwks_uri"], timeout=_HTTP_TIMEOUT)
            resp.raise_for_status()
            self._jwks = resp.json()
        return self._jwks

    def authorization_url(
        self,
        *,
        state: str,
        nonce: str,
        code_challenge: str,
        redirect_uri: str,
    ) -> str:
        params: dict[str, str] = {
            "response_type": "code",
            "client_id": self.client_id,
            "redirect_uri": redirect_uri,
            "scope": self.scope,
            "state": state,
            "nonce": nonce,
            "code_challenge": code_challenge,
            "code_challenge_method": "S256",
        }
        params.update(self._extra_auth_params())
        return f"{self._get_discovery()['authorization_endpoint']}?{urlencode(params)}"

    def _fetch_token(
        self,
        *,
        code: str,
        code_verifier: str,
        redirect_uri: str,
    ) -> dict[str, Any]:
        """POST the auth code to the token endpoint. Returns the raw token dict.

        Isolated so tests can stub it without involving Authlib's HTTP stack.
        """
        client = OAuth2Session(
            client_id=self.client_id,
            client_secret=self.client_secret,
        )
        try:
            token = client.fetch_token(
                self._get_discovery()["token_endpoint"],
                code=code,
                code_verifier=code_verifier,
                redirect_uri=redirect_uri,
                grant_type="authorization_code",
            )
        except Exception as e:
            raise IdentityFetchError(f"token exchange failed: {e}") from e
        return dict(token)

    def exchange_code(
        self,
        *,
        code: str,
        code_verifier: str,
        redirect_uri: str,
        expected_nonce: str | None,
    ) -> ProviderIdentity:
        token = self._fetch_token(
            code=code, code_verifier=code_verifier, redirect_uri=redirect_uri
        )
        id_token = token.get("id_token")
        if not id_token:
            raise IdentityFetchError("no id_token in token response")

        try:
            # joserfc's KeySet.import_key_set is typed as accepting its own
            # KeySetSerialization TypedDict, but operationally it accepts any
            # standard JWKS dict. Mypy 2.x widened the stub; 1.x still
            # complains. The dual code suppresses both directions.
            key_set = KeySet.import_key_set(self._get_jwks())  # type: ignore[arg-type, unused-ignore]
            decoded = jwt.decode(id_token, key_set)
            claims = decoded.claims
        except JoseError as e:
            raise IdentityFetchError(f"id_token validation failed: {e}") from e

        issuer = self._get_discovery().get("issuer")
        if issuer and claims.get("iss") != issuer:
            raise IdentityFetchError(
                f"id_token issuer mismatch: expected {issuer!r}, got {claims.get('iss')!r}"
            )
        if not _audience_matches(claims.get("aud"), self.client_id):
            raise IdentityFetchError(f"id_token audience mismatch: {claims.get('aud')!r}")
        if expected_nonce is not None and claims.get("nonce") != expected_nonce:
            raise IdentityFetchError("id_token nonce mismatch")

        exp = claims.get("exp")
        if isinstance(exp, (int, float)) and exp < int(time.time()) - 60:
            # 60s clock skew tolerance.
            raise IdentityFetchError("id_token expired")

        sub = claims.get("sub")
        if not sub:
            raise IdentityFetchError("id_token missing 'sub' claim")

        return ProviderIdentity(
            slot=self.slot,
            subject=str(sub),
            email=claims.get("email"),
            email_verified=bool(claims.get("email_verified")),
            name=claims.get("name"),
            picture=claims.get("picture"),
        )


class GoogleProvider(_OIDCProviderBase):
    slot = "google"
    type = "google"
    discovery_url = GOOGLE_DISCOVERY_URL

    def __init__(self, config: GoogleConfig, *, http: HttpClient | None = None) -> None:
        super().__init__(http=http)
        self.config = config
        self.label = config.label
        self.client_id = config.client_id
        self.client_secret = config.client_secret

    def _extra_auth_params(self) -> dict[str, str]:
        params = {"prompt": "select_account"}
        # When exactly one domain is allowlisted, pass it as ``hd`` so Google
        # pre-selects the right account. This is a UX hint only — the real
        # enforcement happens in ``check_allowlist``.
        if len(self.config.allowed_domains) == 1:
            params["hd"] = self.config.allowed_domains[0]
        return params

    def check_allowlist(self, identity: ProviderIdentity) -> None:
        if not self.config.allowed_domains:
            return
        if not identity.email or not identity.email_verified:
            raise AllowlistDenied("verified email required for domain check")
        domain = _email_domain(identity.email)
        allowed = {d.lower() for d in self.config.allowed_domains}
        if domain not in allowed:
            raise AllowlistDenied(f"email domain {domain!r} is not in the allowed domains list")


class GenericOIDCProvider(_OIDCProviderBase):
    type = "oidc"

    def __init__(self, config: OIDCConfig, *, http: HttpClient | None = None) -> None:
        super().__init__(http=http)
        self.config = config
        self.slot = config.slot
        self.label = config.label or config.slot.title()
        self.client_id = config.client_id
        self.client_secret = config.client_secret
        self.discovery_url = config.discovery_url

    def check_allowlist(self, identity: ProviderIdentity) -> None:
        if not self.config.allowed_domains:
            return
        if not identity.email or not identity.email_verified:
            raise AllowlistDenied("verified email required for domain check")
        domain = _email_domain(identity.email)
        allowed = {d.lower() for d in self.config.allowed_domains}
        if domain not in allowed:
            raise AllowlistDenied(f"email domain {domain!r} is not in the allowed domains list")


# ── GitHub (OAuth2-only, no OIDC) ──────────────────────────────────────


class GitHubProvider:
    slot = "github"
    type = "github"
    scope = "read:user user:email"

    def __init__(self, config: GitHubConfig, *, http: HttpClient | None = None) -> None:
        self.config = config
        self.label = config.label
        self._http = http or requests.Session()

    def authorization_url(
        self,
        *,
        state: str,
        nonce: str,
        code_challenge: str,
        redirect_uri: str,
    ) -> str:
        # GitHub does not implement OIDC: ``nonce`` is unused. PKCE is honoured
        # — GitHub added support for it in 2023.
        params = {
            "client_id": self.config.client_id,
            "redirect_uri": redirect_uri,
            "scope": self.scope,
            "state": state,
            "code_challenge": code_challenge,
            "code_challenge_method": "S256",
            "allow_signup": "false",
        }
        # Request read:org scope when allowlist is configured, so the
        # membership endpoint returns reliable results.
        if self.config.allowed_orgs:
            params["scope"] = self.scope + " read:org"
        return f"{GITHUB_AUTH_URL}?{urlencode(params)}"

    def _fetch_token(
        self,
        *,
        code: str,
        code_verifier: str,
        redirect_uri: str,
    ) -> dict[str, Any]:
        client = OAuth2Session(
            client_id=self.config.client_id,
            client_secret=self.config.client_secret,
        )
        try:
            token = client.fetch_token(
                GITHUB_TOKEN_URL,
                code=code,
                code_verifier=code_verifier,
                redirect_uri=redirect_uri,
                grant_type="authorization_code",
                # GitHub returns form-encoded by default; ask for JSON.
                headers={"Accept": "application/json"},
            )
        except Exception as e:
            raise IdentityFetchError(f"token exchange failed: {e}") from e
        return dict(token)

    def _api_get(self, path: str, access_token: str) -> Any:
        resp = self._http.get(
            f"{GITHUB_API_BASE}{path}",
            headers={
                "Authorization": f"Bearer {access_token}",
                "Accept": "application/vnd.github+json",
                "X-GitHub-Api-Version": "2022-11-28",
            },
            timeout=_HTTP_TIMEOUT,
        )
        if resp.status_code >= 400 and resp.status_code != 404:
            raise IdentityFetchError(
                f"GitHub API {path} returned {resp.status_code}: {resp.text[:200]}"
            )
        return resp

    def exchange_code(
        self,
        *,
        code: str,
        code_verifier: str,
        redirect_uri: str,
        expected_nonce: str | None,
    ) -> ProviderIdentity:
        token = self._fetch_token(
            code=code, code_verifier=code_verifier, redirect_uri=redirect_uri
        )
        access_token = token.get("access_token")
        if not access_token:
            raise IdentityFetchError("no access_token in token response")

        user_resp = self._api_get("/user", access_token)
        if user_resp.status_code != 200:
            raise IdentityFetchError(f"GET /user failed: {user_resp.status_code}")
        user = user_resp.json()
        gh_id = user.get("id")
        login = user.get("login")
        if gh_id is None or not login:
            raise IdentityFetchError("GitHub /user response missing 'id' or 'login'")

        primary_email, verified = self._primary_email(access_token)

        # Org membership requires the access token, so we enforce it here
        # rather than in ``check_allowlist`` (which is a no-op for GitHub).
        # Any denial raises :class:`AllowlistDenied` straight through.
        self._verify_org_membership(access_token, str(login))

        return ProviderIdentity(
            slot=self.slot,
            subject=str(gh_id),
            email=primary_email,
            email_verified=verified,
            name=user.get("name") or user.get("login"),
            picture=user.get("avatar_url"),
        )

    def _primary_email(self, access_token: str) -> tuple[str | None, bool]:
        """Return ``(primary_verified_email_or_None, verified_flag)``.

        Falls back to ``None`` if no verified primary exists. We never trust
        an unverified email for any access decision.
        """
        resp = self._api_get("/user/emails", access_token)
        if resp.status_code != 200:
            return None, False
        for entry in resp.json():
            if entry.get("primary") and entry.get("verified"):
                return entry.get("email"), True
        return None, False

    def _verify_org_membership(self, access_token: str, login: str) -> None:
        if not self.config.allowed_orgs:
            return
        for org in self.config.allowed_orgs:
            resp = self._http.get(
                f"{GITHUB_API_BASE}/orgs/{org}/members/{login}",
                headers={
                    "Authorization": f"Bearer {access_token}",
                    "Accept": "application/vnd.github+json",
                    "X-GitHub-Api-Version": "2022-11-28",
                },
                timeout=_HTTP_TIMEOUT,
            )
            if resp.status_code == 204:
                return
            if resp.status_code not in (302, 404):
                raise IdentityFetchError(f"GitHub org membership check failed: {resp.status_code}")
        raise AllowlistDenied(
            f"user is not a member of any allowed GitHub org "
            f"({', '.join(self.config.allowed_orgs)})"
        )

    def check_allowlist(self, identity: ProviderIdentity) -> None:
        """No-op — GitHub's org check happens inside :meth:`exchange_code`.

        Required by the :class:`OAuthProvider` protocol for interface symmetry.
        """
        return

File: sdks/node/src/dashboard/auth/oauth/providers.ts
// Concrete providers: Google, GitHub, and generic OIDC. `exchangeCode`
// (network IO + claim normalisation) is split from `checkAllowlist`
// (pure-data permission check) so tests can drive either in isolation.
// HTTP goes through an injectable `fetch` so tests never hit the network.

import type { GitHubConfig, GoogleConfig, OidcConfig } from "./config";
import {
  AllowlistDenied,
  IdentityFetchError,
  type OAuthProvider,
  type ProviderIdentity,
} from "./identity";
import { validateIdToken } from "./idToken";

export const GOOGLE_DISCOVERY_URL = "https://accounts.google.com/.well-known/openid-configuration";
const GITHUB_AUTH_URL = "https://github.com/login/oauth/authorize";
const GITHUB_TOKEN_URL = "https://github.com/login/oauth/access_token";
const GITHUB_API_BASE = "https://api.github.com";

const HTTP_TIMEOUT_MS = 10_000;

export type FetchLike = typeof fetch;

interface DiscoveryDocument {
  issuer?: string;
  authorization_endpoint?: string;
  token_endpoint?: string;
  jwks_uri?: string;
}

function emailDomain(email: string | null): string | null {
  if (!email?.includes("@")) {
    return null;
  }
  return (email.split("@").pop() ?? "").toLowerCase();
}

function domainAllowlistCheck(identity: ProviderIdentity, allowedDomains: string[]): void {
  if (allowedDomains.length === 0) {
    return;
  }
  if (!identity.email || !identity.emailVerified) {
    throw new AllowlistDenied("verified email required for domain check");
  }
  const domain = emailDomain(identity.email);
  if (!allowedDomains.some((d) => d.toLowerCase() === domain)) {
    throw new AllowlistDenied(`email domain '${domain}' is not in the allowed domains list`);
  }
}

/**
 * Discovered endpoints are attacker-influenceable via a misconfigured or
 * compromised discovery document; require https (plain http only for
 * local development hosts) before sending codes or the client secret.
 */
function requireSecureEndpoint(raw: string | undefined, field: string): string {
  if (!raw) {
    throw new IdentityFetchError(`OIDC discovery document has no ${field}`);
  }
  let parsed: URL;
  try {
    parsed = new URL(raw);
  } catch {
    throw new IdentityFetchError(`OIDC discovery ${field} is not a valid URL`);
  }
  const isLocal = ["localhost", "127.0.0.1", "::1", "[::1]"].includes(parsed.hostname);
  if (parsed.protocol !== "https:" && !(parsed.protocol === "http:" && isLocal)) {
    throw new IdentityFetchError(`OIDC discovery ${field} must use https: ${raw}`);
  }
  return raw;
}

async function fetchJson(http: FetchLike, url: string, what: string): Promise<unknown> {
  let response: Response;
  try {
    response = await http(url, { signal: AbortSignal.timeout(HTTP_TIMEOUT_MS) });
  } catch (error) {
    throw new IdentityFetchError(`${what} request failed: ${String(error)}`);
  }
  if (!response.ok) {
    throw new IdentityFetchError(`${what} returned ${response.status}`);
  }
  try {
    return await response.json();
  } catch {
    throw new IdentityFetchError(`${what} returned invalid JSON`);
  }
}

/** POST an authorization code to a token endpoint (client_secret_post). */
async function fetchToken(
  http: FetchLike,
  tokenEndpoint: string,
  params: {
    clientId: string;
    clientSecret: string;
    code: string;
    codeVerifier: string;
    redirectUri: string;
  },
): Promise<Record<string, unknown>> {
  const body = new URLSearchParams({
    grant_type: "authorization_code",
    client_id: params.clientId,
    client_secret: params.clientSecret,
    code: params.code,
    code_verifier: params.codeVerifier,
    redirect_uri: params.redirectUri,
  });
  let response: Response;
  try {
    response = await http(tokenEndpoint, {
      method: "POST",
      headers: {
        "content-type": "application/x-www-form-urlencoded",
        accept: "application/json",
      },
      body: body.toString(),
      signal: AbortSignal.timeout(HTTP_TIMEOUT_MS),
    });
  } catch (error) {
    throw new IdentityFetchError(`token exchange failed: ${String(error)}`);
  }
  if (!response.ok) {
    throw new IdentityFetchError(`token exchange failed: HTTP ${response.status}`);
  }
  try {
    return (await response.json()) as Record<string, unknown>;
  } catch {
    throw new IdentityFetchError("token endpoint returned invalid JSON");
  }
}

// ── OIDC (shared machinery for Google + generic OIDC) ───────────────────

abstract class OidcProviderBase implements OAuthProvider {
  abstract readonly slot: string;
  abstract readonly label: string;
  abstract readonly type: string;
  protected abstract readonly clientId: string;
  protected abstract readonly clientSecret: string;
  protected abstract readonly discoveryUrl: string;
  protected readonly scope: string = "openid email profile";

  private discovery?: DiscoveryDocument;
  private jwks?: { keys?: Array<Record<string, unknown>> };

  constructor(protected readonly http: FetchLike = fetch) {}

  protected extraAuthParams(): Record<string, string> {
    return {};
  }

  abstract checkAllowlist(identity: ProviderIdentity): void;

  private async getDiscovery(): Promise<DiscoveryDocument> {
    this.discovery ??= (await fetchJson(
      this.http,
      this.discoveryUrl,
      "OIDC discovery",
    )) as DiscoveryDocument;
    return this.discovery;
  }

  private async getJwks(): Promise<{ keys?: Array<Record<string, unknown>> }> {
    if (!this.jwks) {
      const discovery = await this.getDiscovery();
      const jwksUri = requireSecureEndpoint(discovery.jwks_uri, "jwks_uri");
      this.jwks = (await fetchJson(this.http, jwksUri, "JWKS fetch")) as {
        keys?: Array<Record<string, unknown>>;
      };
    }
    return this.jwks;
  }

  async authorizationUrl(params: {
    state: string;
    nonce: string;
    codeChallenge: string;
    redirectUri: string;
  }): Promise<string> {
    const discovery = await this.getDiscovery();
    const authorizationEndpoint = requireSecureEndpoint(
      discovery.authorization_endpoint,
      "authorization_endpoint",
    );
    const query = new URLSearchParams({
      response_type: "code",
      client_id: this.clientId,
      redirect_uri: params.redirectUri,
      scope: this.scope,
      state: params.state,
      nonce: params.nonce,
      code_challenge: params.codeChallenge,
      code_challenge_method: "S256",
      ...this.extraAuthParams(),
    });
    return `${authorizationEndpoint}?${query.toString()}`;
  }

  async exchangeCode(params: {
    code: string;
    codeVerifier: string;
    redirectUri: string;
    expectedNonce: string | null;
  }): Promise<ProviderIdentity> {
    const discovery = await this.getDiscovery();
    const tokenEndpoint = requireSecureEndpoint(discovery.token_endpoint, "token_endpoint");
    const token = await fetchToken(this.http, tokenEndpoint, {
      clientId: this.clientId,
      clientSecret: this.clientSecret,
      code: params.code,
      codeVerifier: params.codeVerifier,
      redirectUri: params.redirectUri,
    });
    const idToken = token.id_token;
    if (typeof idToken !== "string" || !idToken) {
      throw new IdentityFetchError("no id_token in token response");
    }
    const claims = validateIdToken({
      idToken,
      jwks: await this.getJwks(),
      issuer: discovery.issuer,
      clientId: this.clientId,
      expectedNonce: params.expectedNonce,
    });
    return {
      slot: this.slot,
      subject: String(claims.sub),
      email: typeof claims.email === "string" ? claims.email : null,
      emailVerified: claims.email_verified === true,
      name: typeof claims.name === "string" ? claims.name : null,
      picture: typeof claims.picture === "string" ? claims.picture : null,
    };
  }
}

export class GoogleProvider extends OidcProviderBase {
  readonly slot = "google";
  readonly type = "google";
  readonly label: string;
  protected readonly clientId: string;
  protected readonly clientSecret: string;
  protected readonly discoveryUrl = GOOGLE_DISCOVERY_URL;

  constructor(
    private readonly config: GoogleConfig,
    http: FetchLike = fetch,
  ) {
    super(http);
    this.label = config.label;
    this.clientId = config.clientId;
    this.clientSecret = config.clientSecret;
  }

  protected override extraAuthParams(): Record<string, string> {
    const params: Record<string, string> = { prompt: "select_account" };
    // A single allowlisted domain is passed as `hd` so Google pre-selects
    // the right account. UX hint only — enforcement is in checkAllowlist.
    if (this.config.allowedDomains.length === 1 && this.config.allowedDomains[0]) {
      params.hd = this.config.allowedDomains[0];
    }
    return params;
  }

  checkAllowlist(identity: ProviderIdentity): void {
    domainAllowlistCheck(identity, this.config.allowedDomains);
  }
}

export class GenericOidcProvider extends OidcProviderBase {
  readonly slot: string;
  readonly type = "oidc";
  readonly label: string;
  protected readonly clientId: string;
  protected readonly clientSecret: string;
  protected readonly discoveryUrl: string;

  constructor(
    private readonly config: OidcConfig,
    http: FetchLike = fetch,
  ) {
    super(http);
    this.slot = config.slot;
    this.label = config.label || config.slot;
    this.clientId = config.clientId;
    this.clientSecret = config.clientSecret;
    this.discoveryUrl = config.discoveryUrl;
  }

  checkAllowlist(identity: ProviderIdentity): void {
    domainAllowlistCheck(identity, this.config.allowedDomains);
  }
}

// ── GitHub (OAuth2-only, no OIDC) ────────────────────────────────────────

export class GitHubProvider implements OAuthProvider {
  readonly slot = "github";
  readonly type = "github";
  readonly label: string;
  private readonly scope = "read:user user:email";

  constructor(
    private readonly config: GitHubConfig,
    private readonly http: FetchLike = fetch,
  ) {
    this.label = config.label;
  }

  // GitHub does not implement OIDC — `nonce` is unused; PKCE is honoured.
  authorizationUrl(params: {
    state: string;
    nonce: string;
    codeChallenge: string;
    redirectUri: string;
  }): Promise<string> {
    const query = new URLSearchParams({
      client_id: this.config.clientId,
      redirect_uri: params.redirectUri,
      // read:org makes the membership endpoint reliable when allowlisting.
      scope: this.config.allowedOrgs.length > 0 ? `${this.scope} read:org` : this.scope,
      state: params.state,
      code_challenge: params.codeChallenge,
      code_challenge_method: "S256",
      allow_signup: "false",
    });
    return Promise.resolve(`${GITHUB_AUTH_URL}?${query.toString()}`);
  }

  async exchangeCode(params: {
    code: string;
    codeVerifier: string;
    redirectUri: string;
    expectedNonce: string | null;
  }): Promise<ProviderIdentity> {
    const token = await fetchToken(this.http, GITHUB_TOKEN_URL, {
      clientId: this.config.clientId,
      clientSecret: this.config.clientSecret,
      code: params.code,
      codeVerifier: params.codeVerifier,
      redirectUri: params.redirectUri,
    });
    const accessToken = token.access_token;
    if (typeof accessToken !== "string" || !accessToken) {
      throw new IdentityFetchError("no access_token in token response");
    }

    const user = (await this.apiGet("/user", accessToken, "GET /user")) as {
      id?: number;
      login?: string;
      name?: string;
      avatar_url?: string;
    };
    if (user.id === undefined || !user.login) {
      throw new IdentityFetchError("GitHub /user response missing 'id' or 'login'");
    }

    const { email, verified } = await this.primaryEmail(accessToken);
    // Org membership needs the access token, so it is enforced here rather
    // than in checkAllowlist (a no-op for GitHub).
    await this.verifyOrgMembership(accessToken, user.login);

    return {
      slot: this.slot,
      subject: String(user.id),
      email,
      emailVerified: verified,
      name: user.name ?? user.login,
      picture: user.avatar_url ?? null,
    };
  }

  /** No-op — GitHub's org check happens inside exchangeCode. */
  checkAllowlist(_identity: ProviderIdentity): void {}

  private async apiGet(path: string, accessToken: string, what: string): Promise<unknown> {
    const response = await this.apiRequest(path, accessToken);
    if (!response.ok) {
      throw new IdentityFetchError(`${what} failed: ${response.status}`);
    }
    try {
      return await response.json();
    } catch {
      throw new IdentityFetchError(`${what} returned invalid JSON`);
    }
  }

  private async apiRequest(path: string, accessToken: string): Promise<Response> {
    try {
      return await this.http(`${GITHUB_API_BASE}${path}`, {
        headers: {
          authorization: `Bearer ${accessToken}`,
          accept: "application/vnd.github+json",
          "x-github-api-version": "2022-11-28",
        },
        signal: AbortSignal.timeout(HTTP_TIMEOUT_MS),
      });
    } catch (error) {
      throw new IdentityFetchError(`GitHub API ${path} request failed: ${String(error)}`);
    }
  }

  /** The verified primary email, or null — unverified emails are never trusted. */
  private async primaryEmail(
    accessToken: string,
  ): Promise<{ email: string | null; verified: boolean }> {
    const response = await this.apiRequest("/user/emails", accessToken);
    if (!response.ok) {
      return { email: null, verified: false };
    }
    let entries: Array<{ email?: string; primary?: boolean; verified?: boolean }>;
    try {
      entries = (await response.json()) as typeof entries;
    } catch {
      return { email: null, verified: false };
    }
    for (const entry of entries) {
      if (entry.primary && entry.verified && entry.email) {
        return { email: entry.email, verified: true };
      }
    }
    return { email: null, verified: false };
  }

  private async verifyOrgMembership(accessToken: string, login: string): Promise<void> {
    if (this.config.allowedOrgs.length === 0) {
      return;
    }
    for (const org of this.config.allowedOrgs) {
      const response = await this.apiRequest(
        `/orgs/${encodeURIComponent(org)}/members/${encodeURIComponent(login)}`,
        accessToken,
      );
      if (response.status === 204) {
        return;
      }
      if (response.status !== 302 && response.status !== 404) {
        throw new IdentityFetchError(`GitHub org membership check failed: ${response.status}`);
      }
    }
    throw new AllowlistDenied(
      `user is not a member of any allowed GitHub org (${this.config.allowedOrgs.join(", ")})`,
    );
  }
}
